### DISCUSSION TITLE
Why values set by `setValue` are not get on `getValues` and how to get it?

### DISCUSSION THREAD

**Discussion by ghost** [react-hook-form/react-hook-form#3896] [Q&A] [answered]
I have set the values of a field using `methods.setValue('name', value, {shouldValidate: true, shouldDirty: true })` and get them using `methods.getValues()`, but only the fields registered through `register` are get by `methods.getValues()`.

At the beginning of the component, I'm using `methods.reset(defaultValuesFromNetwork)` and besides it the usge is like the examples on the doc.

Version 6.11.4.

**Comment by bluebill1049** [comment, marked as answer]
> An optimized helper for reading form values. The difference between watch and getValues is that getValues will not trigger re-renders or subscribe to input changes.

getValues is for the read, it doesn't subscribe to the form. You want `watch` or `useWatch`.

**Reply by ghost** [reply]
Yeah, I re-read everything and I saw that I need to set the default values manually for fields that use `setValue`, the problem now its that when I open the form the default values are not rendered, even though I have used `reset`. If I click on another tab (inside the same page) and come back to the form it renders the default values, I guess cause a new render was done. I'm already passing the values to each subcomponent through watch:

![image](https://user-images.githubusercontent.com/19755014/104533123-882b8e80-55f0-11eb-9936-64b4c7328f59.png)

![image](https://user-images.githubusercontent.com/19755014/104533662-90d09480-55f1-11eb-992c-de7fd4f85e32.png)

![image](https://user-images.githubusercontent.com/19755014/104533182-a7c2b700-55f0-11eb-87c8-afcdc8a6ed82.png)

**Reply by bluebill1049** [reply]
ideally, you should have useForm per tab, and host your data in a store lib instead rely on useForm to keep all input state. at least that the design for v6 made for. v7 we are addressing this issue better, but still early stage.

**Reply by ghost** [reply]
I'm doing exactly it, each tab is a form and in the submit of each tab the data is placed in a React state ( :

**Reply by ghost** [reply]
The only thing is that the React Hook Forms don't set the default data that is there, cause I checked on each point if is it being stored, received / sent on request, and persisted in the state

**Reply by bluebill1049** [reply]
can you share a simple codesandbox?

**Reply by ghost** [reply]
Hmmm, a real example no, cause the architecture has a lot of pieces, but the general idea is that:

1. Create a form `useForm` passing the default values, which come from an API and is stored in a React state

2. When the component is rendered call `reset`

3. Create a `useWatch` that receive the same default values as the `useForm`

4. Use `useWatch` to pass the default value for each field (in the print above a bool field)

5. Inside the field use Control and set the default value in the input itself and each time it changes convert the value from bool string to bool inside the `useForm` through `setValue`

6. When the form is submitted get the data through parameter on `onSubmit` and set it in a React state has an effect that listens for changes and when all tabs are done it sends the data

7. After the resource is created we get it from the API and set it in the default data React state (the data is propagated but it is not shown in the interface until I kinda do a new render by choosing another tab and then come back)

**Reply by ghost** [reply]
Is there some way to force an update on React Hook side? Cause I tried the options that the documentation gives me, but... a last hope [?](https://www.youtube.com/watch?v=D1Nq086QB1Q)

**Reply by bluebill1049** [reply]
if you need a force update. `const [_, forceUpdate] = React.useState({})` `forceUpdate({})`

**Reply by ghost** [reply]
Fixed!

> Force the update will do nothing, cause [states are async](https://reactjs.org/docs/faq-state.html#why-is-setstate-giving-me-the-wrong-value) in React.

Errors:

1. Cause the state is asynchronous you have issues with network data and render 

2. Set `defaultValues` in `useWatch` was causing the data to not being loaded, even though it was also set in `useForm`

Solutions:

1. Tried to use `useLayoutEffect` but didn't work and the old (ugly) `loading` variable and `if(loading) return <pre>Loading, never come back</pre>` which worked

2. I removed the `defaultValues` from `useWatch` and instead of using the hook I used the `watch` from `methods.watch` which comes from `useForm` directly

**Reply by bluebill1049** [reply]
i see, we are fixing watch/useWatch defaultValue, it shouldn't be a fallback for undefined, in v7, it will just be the default value for the initial render.

**Reply by ghost** [reply]
That's awesome, expecting the new version.
